src/Controller/ApiController.php

```diff
@@ -225,6 +227,35 @@ public function editPackageAction(Request $request, Package $package, ValidatorI
         return new JsonResponse(['status' => 'success'], 200);
     }
 
+    #[Route(path: '/api/packages/{package}/delete-version/{version}', name: 'api_delete_package_version', requirements: ['package' => '[A-Za-z0-9_.-]+/[A-Za-z0-9_.-]+?'], defaults: ['_format' => 'json'], methods: ['POST'])]
+    public function deletePackageVersionAction(Request $request, #[MapEntity(mapping: ['package' => 'name'])] Package $package, string $version, StatsDClient $statsd): JsonResponse
+    {
+        $user = $this->findUser($request);
+        if (!$user) {
+            return new JsonResponse(['status' => 'error', 'message' => 'Missing or invalid username/apiToken in request'], 406);
+        }
+        if (!$package->getMaintainers()->contains($user)) {
+            return new JsonResponse(['status' => 'error', 'message' => 'You are not allowed to delete this package version'], 403);
+        }
+
+        $statsd->increment('delete_package_version_api');
+
+        $packageVersion = $package->getVersions()->findFirst(static function (int $index, Version $innerVersion) use ($version) {
+            return $innerVersion->getVersion() === $version;
+        });
+
+        if (!$packageVersion) {
+            return new JsonResponse(['status' => 'error', 'message' => 'Version not found'], 404);
+        }
+
+        $repo = $this->getEM()->getRepository(Version::class);
+        $repo->remove($packageVersion);
```

Comment:
this should be wrapped in `$em->transactional()` (which calls `flush()` automatically at the end) or `$em->getConnection()->transactional()` (keeping an explicit flush of the entity manager inside the closure) to ensure that the removal is done in a transactional way.